ships/0031-shipwright-trigger.md

```diff
@@ -0,0 +1,383 @@
+
+# Proposal
+
+Introduce a new project name Shipwright Trigger (`shipwright-io/trigger`) to based on the rules defined on the Build resources, listen to WebHook events coming from Git service providers, incorporate Shipwright into Tekton via `CustomTask` (Tekton Run resource), and watch over Tekton Pipeline resources in order to trigger `BuildRun` resources accordingly.
+
+## API Changes
+
+We will modify the `Build` CRD to allow users to define how to trigger it. The API will look like the following:
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+spec:
+  source:
+    url: https://github.com/otaviof/nodejs-ex.git
+    credentials:
+      name: webhook-secret
+  trigger:
+    when:
+      - name: for push directly on the main branch
+        type: Push
+        branches:
+          - main
+      - name: for pull-requests on the main branch
+        type: PullRequest
+        branches:
+          - main
+      - name: watching over the base-image
+        type: Image
+        images:
+          - ghcr.io/some/base-image:latest
+      - name: when a Tekton Pipeline is "Succeeded"
+        type: Pipeline
+        objectRef:
+          status:
+            - Succeeded
+          selector:
+            label: value
+```
+
+The trigger is defined directly on the `Build` resource, and when the attribute is empty it means the given `Build` resource is not subject to triggers, as expected. The new `.spec.trigger` attribute is structured as follows:
+
+- **`.source.credentials.name`**: local reference to a Secret object, containing pre-defined keys to share an [secret-token][githubWebHookSecretToken] between Git service providers and local `Build` instances, reusing the exiting Secret for the Git repository authentication;
+- **`.when[]`**: slice of scenarios to describe *when* the `Build` will be activated;
+- **`.when[].name`**: short description of what the scenario entails;
+- **`.when[].type`**: the type of scenario, namely: `Push`, `PullRequest`, `Image` and `Pipeline`;
+- **`.when[].branches[]`**: slice of branch names for the current repository, branch names where a Build should be activated in case of `Push` or `PullRequest` event. When not informed, it will adopt the attribute `.spec.source.revision` instead;
+- **`.when[].images[]`**: slice of image names to watch, when those are updated a new Build is activated;
+- **`.when[].objectRef`**: used in conjunction with `Pipeline`(type), and describes a local Tekton Pipeline instance and its current status;
+- **`.when[].objectRef.name`**: when informed, the Pipeline is directly search by its name;
+- **`.when[].objectRef.status[]`**: slice of Tekton Pipeline [status names][tektonPipelineRunStatus] where a new Build should be activated. For instance *"Succeeded"*;
+- **`.when[].objectRef.selector{}`**: a map of label key-value entries, a alternative to the `.objectRef.name` attribute, the Tekton Pipeline resource will be identified by label matching;
```

Comment:
I recommend that we have separate API objects for each trigger type:

- `.when[].pullRequest.branches[]`
- `.when[].image.images[]`
- `.when[].pipeline.name`

This lets us employ type discriminators and add validation to the API (ex - the `Git` trigger type should not have any `objectRef` values)

I won't block merge on this - we can discuss in further detail when the full API is implemented.

Reply:
I tend to agree. We do this now with volumes where we inherit the Kubernetes VolumeSource. And imo that's also what we should go to with `spec.sources` in our future API evolvement.

Reply:
Thanks for the recommendation, @adambkaplan. I've adjusted the EP accordingly, and as well the prototype changes. Please consider.